### PR TITLE
[Data] Fix '_unwrap_protocol' for Windows systems

### PR DESCRIPTION
**Pull request by jbedorf**
Signed-off-by: Jeroen Bédorf <jeroen@bedorf.net>

## Why are these changes needed?

The `_unwrap_protocol` method uses the `urllib.parse.urlparse` library function to split out the path and protocol. On Windows however this function returns the path with a `/` added before the drive letter. This type of path can't be used by any other functions. The solution is to strip the `/`. The logic is similar to what is used in the `pip` package, see [here](https://github.com/pypa/pip/blob/22.3.1/src/pip/_internal/utils/urls.py#L49).

Example script that shows the issue. Without this fix the below will not work.

```
import pandas as pd
from ray.data.datasource.file_based_datasource import _unwrap_protocol
import os

import ray
import pyarrow as pa
import pyarrow.parquet as pq

data_path = "file:///C://tmp"

df1 = pd.DataFrame({"one": [1, 2, 3], "two": ["a", "b", "c"]})
table = pa.Table.from_pandas(df1)
setup_data_path = _unwrap_protocol(data_path)
print("Setup data path: ", setup_data_path)
path1 = os.path.join(setup_data_path, "test1.parquet")
pq.write_table(table, path1, filesystem=None)

```
Output before: `/C://tmp`

Output after: `C://tmp`

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by c21** [comment]
Hi @jbedorf - thanks for the fix! Seeing a bunch of CI tests failure, and it seems not related to this PR. Could you rebase to latest master? Thanks!

**Comment by jbedorf** [comment]
> Hi @jbedorf - thanks for the fix! Seeing a bunch of CI tests failure, and it seems not related to this PR. Could you rebase to latest master? Thanks!

The tests now pass, with the exception of some unrelated tests due to the numpy issues.

**Review by jianoaix — COMMENTED** [review]
Could you add a unit test, with the example you already have? Thanks.

**Comment by jbedorf** [comment]
> Could you add a unit test, with the example you already have? Thanks.

Sure, but where do you want me to put it? Given [this comment](https://github.com/ray-project/ray/blob/master/python/ray/data/datasource/file_based_datasource.py#L570) in the original file it looks like there are no existing tests.

My above example is based on one of the existing dataset tests, but I guess that these have never been run on Windows. When I tried it just now I got tons of errors related to file-system operations. The interesting thing is that they involve the same function that I fixed here. For example just running this test: ` .\python\ray\data\tests\test_dataset_json.py::test_json_read[None-local_path-None]` . Will give an error. 
The error that it gives is that it tries to compare these urls:

``C:/Users/<user>/AppData.../test1.json' != '\\Users\\<user>\\AppData\\...test1.json``

As you can see the right hand side of the comparison has the drive letter cut off. That's because of the following:

On Linux:
```
>>> urllib.parse.urlparse("\tmp\test")
ParseResult(scheme='', netloc='', path='\tmp\test', params='', query='', fragment='')
```
Something similar on Windows
```
>>> urllib.parse.urlparse("C:\\test")
ParseResult(scheme='c', netloc='', path='\\test', params='', query='', fragment='')
```

Because the drive letter (C) is assigned as scheme, the `_unwrap_protocol` discards it. Now that's exactly the issue, there is no protocol given in this path. This works apparently fine on Linux, but not on Windows where the drive letter is seen as the protocol. This matches [with the documentation](https://docs.python.org/3/library/urllib.parse.html#url-parsing) . So basically the function is misused in the tests. It is being applied to standard, local file-system, paths and not on URLs. So the fully fix it either the `_unwrap_protocol` method should not be used as it currently is. Or it should be rewritten in such a way that it detects if it is a local path instead of a URL and then just return the local path instead of what it does now. 

Someone else noticed this two years ago and filed an issue against the Python libraries [here](https://github.com/python/cpython/issues/86381).

**Comment by jbedorf** [comment]
@jianoaix  I've updated the code to take into account the issues I mentioned [here](https://github.com/ray-project/ray/pull/31296#issuecomment-1363688361).

The changes here fix existing tests such as:
* `.\python\ray\data\tests\test_dataset_json.py::test_json_read[None-local_path-None]`
* `.\python\ray\tune\tests\test_tuner_restore.py::test_tuner_restore_num_trials` 

It won't fix all issues with the existing tests in those files, but the other issues are not stemming from the file path resolution. The changes are all behind if statements that only apply them on Windows systems so none of the existing functionality has changed (as seen in the passing of the existing test suite).

**Comment by jbedorf** [comment]
This is still relevant.
@jianoaix  Please help with getting this reviewed, thanks.

**Comment by jbedorf** [comment]
The failing `dataset-tests-arrow` appear to fail for other PRs as well. The test never returns when running.

**Review by jianoaix — COMMENTED** [review]
Apologies for the delay, LTGM

File: python/ray/data/datasource/file_based_datasource.py [hunk 1]
```diff
@@ -726,10 +727,29 @@ def _decode_url(path):
     return urllib.parse.unquote(path)
 
 
+def _is_local_windows_path(uri: str) -> bool:
```
[the remark this answers is not in the document]
Comment (jbedorf) [reply python/ray/data/datasource/file_based_datasource.py:730]: Sure, but do you want the ray.data package to have a dependency on the ray.air package? Or do you have another location in mind to put this function?

**Comment by richardliaw** [comment]
@jianoaix @c21 is this ready to merge?

File: python/ray/data/datasource/file_based_datasource.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jianoaix) [reply python/ray/data/datasource/file_based_datasource.py:730]: Yes, ray.data already has dependency on ray.air, so that looks fine to me.

**Comment by jianoaix** [comment]
> @jianoaix @c21 is this ready to merge?

Just one comment to address.

**Review by jianoaix — APPROVED** [review]
Thank you again for you patience and contribution!

**Comment by jbedorf** [comment]
@jianoaix Just a ping to get this merged, given that it's approved.

**Comment by c21** [comment]
Sorry for delay, @jbedorf - could you rebase to latest master? After CI is passed, I will merge.

**Comment by c21** [comment]
Merging to master, thanks @jbedorf!

**Comment by jbedorf** [comment]
Thanks @c21 !

**Pull request merged**